### ISSUE TITLE
`torch.compile` not working with lora or xformers

### ISSUE THREAD

**Issue by sunhs** [huggingface/diffusers#3660]
### Describe the bug

I'm experiencing some errors when using `torch.compile` with `UNet2DConditionModel`, possibly caused by either lora or xformers, which I can't decide for now.

I did some tests trying to identify the problem but failed. So I'm posting my test results here, hopefully serving as some useful clues.

I'm not sure whether I should create the issue at `pytorch` or `xformers`. I will if it's more appropriate.

I'm also not sure whether xformers is needed as for pytorch2, since `sdpa` seems to have included that natively.

## test cases 1

```
from diffusers import UNet2DConditionModel
import torch


def test(with_lora: bool, with_xformers: bool, test_attn: bool):
    device = torch.device("cuda:0")
    dtype = torch.half

    model = UNet2DConditionModel.from_pretrained(
        "runwayml/stable-diffusion-v1-5", torch_dtype=dtype, subfolder="unet"
    )

    if with_lora:
        model.load_attn_procs("pcuenq/pokemon-lora")
    if with_xformers:
        model.enable_xformers_memory_efficient_attention()
    print({type(attn_proc) for attn_proc in model.attn_processors.values()})
    if test_attn:
        model = model.down_blocks[0].attentions[0].transformer_blocks[0].attn1

    model = model.to(device)
    model = torch.compile(model)

    if not test_attn:
        latent = torch.randn(1, 4, 64, 64).to(device, dtype)
        timestep = torch.randn(1).to(device, dtype)
        cond = torch.randn(1, 227, 768).to(device, dtype)
        print(model(latent, timestep, cond))
    else:
        sample = torch.randn(1, 320, 64, 64).half().cuda()
        print(model(sample.permute(0, 2, 3, 1).reshape(sample.shape[0], sample.shape[2] * sample.shape[3], sample.shape[1])))


# 1. unet + xformers:
# (32ea21) good
# (9a45d7) good with XFormersAttnProcessor
# test(with_lora=False, with_xformers=True, test_attn=False)

# 2. unet + lora:
# (32ea21) NaN output, So it seems to imply that the problem is caused by lora
# (9a45d7) good with LoRAAttnProcessor2_0, but what causes the difference with LoRAAttnProcessor
# test(with_lora=True, with_xformers=False, test_attn=False)

# 3. unet + lora + xformers:
# (32ea21) syntax error from dynamo, see log_syntax_error.txt. Expected as in #2
# (9a45d7) still syntax error with LoRAXFormersAttnProcessor
# test(with_lora=True, with_xformers=True, test_attn=False)

# 4. attn + lora + xformers:
# (32ea21) error no_dispatcher, see log_no_dispatch.txt
# (9a45d7) still error no_dispatcher with LoRAXFormersAttnProcessor
# test(with_lora=True, with_xformers=True, test_attn=True)

# 5. attn + lora:
# (32ea21) good. WHY? Conflicts with #2 #4
# (9a45d7) good with LoRAAttnProcessor2_0
test(with_lora=True, with_xformers=False, test_attn=True)
```

## test cases 2

```
from diffusers.models.attention_processor import Attention, LoRAAttnProcessor
import torch


def test_manual_attn_with_lora(with_xformers: bool):
    attn = Attention(
        320,
        heads=1,
        dim_head=320,
        bias=True,
        upcast_softmax=True,
        norm_num_groups=32,
        residual_connection=True,
        _from_deprecated_attn_block=True,
    )
    proc = LoRAAttnProcessor(hidden_size=320, cross_attention_dim=320)
    attn.set_processor(proc)

    if with_xformers:
        attn.set_use_memory_efficient_attention_xformers(True)

    attn = attn.half().cuda()
    attn = torch.compile(attn)
    data = torch.randn(1, 4096, 320).half().cuda()
    print(attn(data))


# 6. good
# Conflicts with #2
# test_manual_attn_with_lora(with_xformers=False)

# 7. good again
# Conflicts with #3 #4
# test_manual_attn_with_lora(with_xformers=True)
```

## test cases 3

```
import torch, xformers.ops
import torch._dynamo

torch._dynamo.config.suppress_errors = True


op = torch.compile(xformers.ops.memory_efficient_attention)

q = torch.randn(8, 4096, 40).half().cuda()
k = torch.randn(8, 4096, 40).half().cuda()
v = torch.randn(8, 4096, 40).half().cuda()

# error no_dispatch
# Conflicts with #1
op(q, k, v, scale=0.15811388300841897)
```

### Logs

[log_syntax_error.txt](https://github.com/huggingface/diffusers/files/11642097/log_syntax_error.txt)

[log_no_dispatch.txt](https://github.com/huggingface/diffusers/files/11642098/log_no_dispatch.txt)

### System Info

- `diffusers` version: 0.17.0.dev0, master branch at [32ea21](https://github.com/huggingface/diffusers/tree/32ea2142c056fae722b0cabaa799697a861cd039)
- Platform: Linux-4.9.70-040970-generic-x86_64-with-glibc2.31
- Python version: 3.10.11
- PyTorch version (GPU?): 2.0.1 (True), py3.10_cuda11.8_cudnn8.7.0_0
- Huggingface_hub version: 0.13.3
- Transformers version: 4.29.0
- Accelerate version: 0.19.0
- xFormers version: 0.0.20
- Using GPU in script?: yes
- Using distributed or parallel set-up in script?: no

**Comment by patrickvonplaten** [comment]
I think we have indeed not tested this one yet. @sayakpaul do you think it'd be worth allowing LoRA + xformers in `diffusers`?

**Comment by sayakpaul** [comment]
> @sayakpaul do you think it'd be worth allowing LoRA + xformers in diffusers?

It's supported. You can do the following:

```
from huggingface_hub.repocard import RepoCard
from diffusers import StableDiffusionPipeline
import torch

lora_model_id = "patrickvonplaten/lora_dreambooth_dog_example"
card = RepoCard.load(lora_model_id)
base_model_id = card.data.to_dict()["base_model"]

pipe = StableDiffusionPipeline.from_pretrained(base_model_id, torch_dtype=torch.float16)
pipe = pipe.to("cuda")
pipe.load_lora_weights(lora_model_id)
pipe.enable_xformers_memory_efficient_attention()

image = pipe("A picture of a sks dog in a bucket", num_inference_steps=25).images[0]
```

Not sure if that is what if that is what you mean.

**Comment by patrickvonplaten** [comment]
Sorry, I meant `torch.compile`. The following probs doesn't work yet no? 

```py
from huggingface_hub.repocard import RepoCard
from diffusers import StableDiffusionPipeline
import torch

lora_model_id = "patrickvonplaten/lora_dreambooth_dog_example"
card = RepoCard.load(lora_model_id)
base_model_id = card.data.to_dict()["base_model"]

pipe = StableDiffusionPipeline.from_pretrained(base_model_id, torch_dtype=torch.float16)
pipe = pipe.to("cuda")
pipe.load_lora_weights(lora_model_id)
pipe.unet = torch.compile(pipe.unet)

image = pipe("A picture of a sks dog in a bucket", num_inference_steps=25).images[0]
```

**Comment by sayakpaul** [comment]
On my list. Will provide an update soon.

**Comment by Teoge** [comment]
Any updates? I have similar problems.

**Comment by sayakpaul** [comment]
I just tested @patrickvonplaten's code from https://github.com/huggingface/diffusers/issues/3660#issuecomment-1576834453:

https://colab.research.google.com/gist/sayakpaul/81aaf033df8fd7630a6610dc9947e73c/scratchpad.ipynb

It seems to work. 

The first time after the compiled pipeline is called, it takes a minute. And then it's blazing fast!

**Comment by sunhs** [comment]
Thanks a lot. I tested `test cases 1` again with master branch [9a45d7](https://github.com/huggingface/diffusers/commit/9a45d7fb768d0ee55303cc888667f3b2d9f4c273), and updated the results in the first comment.

It shows that `torch.compile` is working now with `LoRAAttnProcessor2_0`, as shown in `test cases 1 #2`. But `test cases 1 #3` and `test cases 1 #4` remains the same.

IMO this could be a long term problem, as `torch.compile` itself is still in rapid development, where `diffusers` could do nothing.

I'm still curious that, with `sdpa`, is it sufficient to abandon `xformers` ?

**Comment by sayakpaul** [comment]
Hi @sunhs. It's a bit hard to follow from the test cases you mentioned. If you could minimize the snippets where you feel something's off that would be great. 

> I'm still curious that, with sdpa, is it sufficient to abandon xformers ?

I think since SDPA and `torch.compile()` both are in rapid development, the community would still lean on xformers in many cases. This is why we don't restrict users when they want to use xformers.

**Comment by sunhs** [comment]
I have to admit this could be a complicated problem that diffusers could not address on its own. I'll keep track on it and provide information, if any. Closing for now. Thank you guys.

**Comment by sayakpaul** [comment]
Thanks for being understanding and with us. Appreciate that.

**Comment by zengjie617789** [comment]
If used the `torch.compile()` with `load_lora_wights()`, the results seem to be mix up. For example, after the first time I loaded lora wights, and the next time I loaded again, the results is mixed. While not use torch.compile, the results are normal.

**Comment by Dionysus326** [comment]
Is there any update on this after more than one year?
And if not, maybe should address that torch.compile not working with LoRA in the document? like[ this one](https://huggingface.co/docs/diffusers/en/optimization/torch2.0)

**Comment by sayakpaul** [comment]
You should fuse the LoRA before compiling it. See: https://huggingface.co/docs/diffusers/main/en/using-diffusers/merge_loras#torchcompile

**Comment by Dionysus326** [comment]
> You should fuse the LoRA before compiling it. See: https://huggingface.co/docs/diffusers/main/en/using-diffusers/merge_loras#torchcompile

Thanks for the response.
But after following the document, I still got an error
`torch._dynamo.exc.Unsupported: class property UNet2DConditionModel getset_descriptor`

Try to figure out why

**Comment by tolgacangoz** [comment]
With the latest release of xFormers, they say `torch.compile` is supported. [Could you look at it?](https://github.com/facebookresearch/xformers/releases/tag/v0.0.27)

**Comment by Dionysus326** [comment]
I just updated to `xformers==0.0.27` and use the flash attention ops but still could not get it to work. 
It's not needed in my projects tho
So hope anyone has luck with it

**Comment by sayakpaul** [comment]
I am not sure why would you use it given `F.scaled_dot_product_attention()` already provides Flash Attention and is compatible with `torch.compile()`.

**Comment by Dionysus326** [comment]
Oh I see. 
I was using it because when pass `enable_xformers_memory_efficient_attention()` with no argument did not work when I did `torch.compile`
the error message is `torch._dynamo.exc.Unsupported: class property UNet2DConditionModel getset_descriptor`

And I saw this line in the xformers' note :
> fMHA: Added torch.compile support in memory_efficient_attention when passing the flash operator explicitely (eg memory_efficient_attention(..., op=(flash.FwOp, flash.BwOp)))

So I thought specify the ops might help. But it still did not work.
